Update Lisk Elements packages reference - Closes #770 (#790)
LiskHQ/lisk-docs: modules/ROOT/pages/references/lisk-elements/p2p.adoc

P2P node has started!
----

== Architecture

The architecture of the Lisk P2P library is divided into multiple parts (components or classs).
The main parts of P2P library are as follows:

* P2P
* PeerPool
* PeerServer
* PeerBook
* Peer

The responsibilities and interactions of these parts (components or classs) are summarized by the following diagram:

image::p2p.png[Lisk-P2P Architecture]

Before explaining the individual modules, it is helpful to get familiar with the following terminology:

nodeInfo::
Contains the required data for establishing a successful connection, and is updated whenever a peer publishes new (nodeInfo) data.
peerInfo::
Includes shared information which is the nodeInfo of a peer, and also the internal data accumulated on each peer by a node, related to productivity, reputation, rate limit, etc.

=== P2P

> The P2P part is the main entry point of the P2P library that manages PeerPool, PeerBook and PeerServer.

Reviewer: ```suggestion
The P2P class is the main entry point of the P2P library that manages PeerPool, PeerBook and PeerServer.
```

Its responsibilities are to:

* Expose the following actions:
** getConnectedPeers(): Returns a list of all connected peers.
** getDisconnectedPeers(): Returns a list of all disconnected peers.
** applyPenaltyOnPeer(): Applies a penalty score to a specific peer for given peerId.
** applyNodeInfo(): Updates node information.
** getNodeInfo(): Returns node information.
** start(): Starts PeerServer, loads previous peers and triggers discovery through seed nodes if not enough triedPeers.
** stop(): Stops PeerServer, removes all the listeners and calls stop on PeerPool that eventually disconnects all the inbound/outbound connections.

> * Expose actions and emit events bubbling up from  Peer, PeerPool and PeerServer.

Reviewer: ```suggestion
* Expose actions and emit events bubbling up from Peer, PeerPool and PeerServer.
```

* Setup all event listeners and corresponding event handlers.
* Handle getPeers remote procedure calls from peers.

> * Manage CRUD operations on PeerBook that holds and manages all the peerInfo.

Reviewer: ```suggestion
* Manage CRUD operations on PeerBook that holds and manages all the peerInfos of all known peers.
```

> * Load  previous peers and start the discovery through all the seed peers, if not many triedPeers (see section PeerBook).

Reviewer: ```suggestion
* Loads previous peers and does the discovery, in case, if it doesn't have at least 100 triedPeers then it starts the discovery through seed peers provided in the config.
```

* Check for duplicate connections and prevent the node from connecting to itself.
* Check for invalid payloads.
* Perform the handshake check for an incoming connection and create a valid peerInfo object for a peer.

=== Peer

The Peer part is a class which is extended to have specific Outbound and Inbound classes for outgoing and incoming connections respectively.
The outbound class is for outgoing connections which a node makes by creating a socket client, whereas the inbound class is for incoming connections.
Please note that both inbound and outbound peers can be used for 2-way communication, however a node chooses only outbound connections to make a request, and prioritizes outbound peers for sending information.
The responsibilities for both types of connections are to:

* Create or assign a socket for communication on connection.
* Fetch the peer list and status of a peer on connection with outbound peers.
* Impose rate limits on messages and RPC requests, and apply a penalty for misbehaviour.
* Ping inbound peers between 20-30 seconds to check for latency, and also to prove that the connection is alive, otherwise the connection is dropped.

> * Emit all events.

Reviewer: ```suggestion
* Emit all the peer related events including the ones coming from the network.
```

=== PeerBook

The PeerBook part is a directory of peers.
It implements peer buckets for newPeers and triedPeers as described in {}[LIP 0004^].
The newPeers and triedPeers are lists of peer buckets whereby each peer bucket consists of a list of 32 peers, and each peer bucket is chosen based on its IP prefix.

image:peers.png[Peers and buckets] `newPeers` contains peers that either had no connection attempts or failed attempts.

> It consists of  a list of 128 buckets with each bucket having 32 peers.

Reviewer: ```suggestion
It consists of a list of 128 buckets with each bucket having 32 peers.
```
